core/unit_test/TestHostSharedPtr.hpp

```diff
@@ -0,0 +1,247 @@
+/*
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 3.0
+//       Copyright (2020) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#include <impl/Kokkos_HostSharedPtr.hpp>
+
+#include <gtest/gtest.h>
+
+using Kokkos::Experimental::HostSharedPtr;
+using Kokkos::Experimental::UnmanagedPtr;
+
+TEST(TEST_CATEGORY, host_shared_ptr_is_reference_counted) {
+  using T = int;  // default constructible
+  {
+    HostSharedPtr<T> p1;
+    EXPECT_TRUE(p1.is_reference_counted());
+  }
+  {
+    HostSharedPtr<T> p1(nullptr);
+    EXPECT_TRUE(p1.is_reference_counted());
+  }
+  {
+    HostSharedPtr<T> p1(new T());
+    EXPECT_TRUE(p1.is_reference_counted());
+  }
+  {
+    HostSharedPtr<T> p1(new T(), [](T* p) { delete p; });
+    EXPECT_TRUE(p1.is_reference_counted());
+  }
+  {
+    T i;
+    HostSharedPtr<T> p1(&i, [](T*) {});
+    EXPECT_TRUE(p1.is_reference_counted());
+  }
+}
+
+TEST(TEST_CATEGORY, host_shared_ptr_use_count) {
+  using T = int;
+  {
+    HostSharedPtr<T> p1;
+    EXPECT_EQ(p1.use_count(), 1);
+  }
+  {
+    HostSharedPtr<T> p1(nullptr);
+    EXPECT_EQ(p1.use_count(), 1);
+  }
```

Comment:
This is absurd.

Reply:
I would suggest
```diff
diff --git a/core/src/impl/Kokkos_HostSharedPtr.hpp b/core/src/impl/Kokkos_HostSharedPtr.hpp
index 37c4e4da..fea210e8 100644
--- a/core/src/impl/Kokkos_HostSharedPtr.hpp
+++ b/core/src/impl/Kokkos_HostSharedPtr.hpp
@@ -166,7 +167,11 @@ class MaybeReferenceCountedPtr {
 template <class T>
 class HostSharedPtr : public MaybeReferenceCountedPtr<T> {
  public:
-  explicit HostSharedPtr(T* element_ptr = nullptr)
+  HostSharedPtr() noexcept : MaybeReferenceCountedPtr<T>(nullptr) {}
+
+  HostSharedPtr(std::nullptr_t) noexcept : HostSharedPtr() {}
+
+  explicit HostSharedPtr(T* element_ptr)
       : MaybeReferenceCountedPtr<T>(element_ptr, [](T* const t) { delete t; }) {
   }

diff --git a/core/unit_test/TestHostSharedPtr.hpp b/core/unit_test/TestHostSharedPtr.hpp
index eef1feb6..d01d3e5c 100644
--- a/core/unit_test/TestHostSharedPtr.hpp
+++ b/core/unit_test/TestHostSharedPtr.hpp
@@ -78,11 +78,11 @@ TEST(TEST_CATEGORY, host_shared_ptr_use_count) {
   using T = int;
   {
     HostSharedPtr<T> p1;
-    EXPECT_EQ(p1.use_count(), 1);
+    EXPECT_EQ(p1.use_count(), 0);
   }
   {
     HostSharedPtr<T> p1(nullptr);
-    EXPECT_EQ(p1.use_count(), 1);
+    EXPECT_EQ(p1.use_count(), 0);
   }
   {
     HostSharedPtr<T> p1(new T());
```
but that would break
```C++
    {
      HostSharedPtr<T> p1;
      EXPECT_TRUE(p1.is_reference_counted());
    }
    {
      HostSharedPtr<T> p1(nullptr);
      EXPECT_TRUE(p1.is_reference_counted());
    }
```
I would make them return `false` but I believe @dhollman disagree with me on that point

Reply:
I would be fine with these changes.

Reply:
I pushed a corresponding commit. Let's see what @dhollman thinks.

Reply:
LGTM